### PR TITLE
Inverse map functions

### PR DESCRIPTION
**Pull request by austindd**
This PR adds two functions based on [inverseMap](https://hackage.haskell.org/package/relude-0.6.0.0/docs/Relude-Extra-Enum.html#v:inverseMap) from Haskell's `relude` library.

The idea is, given an ordered bounded enum `t` and a function that maps each `t` value to a unique `a` value, we can derive the inverse mapping function from that particular set of `a` values to their corresponding `t` values. Obviously, we can't account for *all* `a` values, so `inverseMap` must return `option(t)`.

This is particularly nice if you already have a `show` function, and want to generate the inverse function to parse strings to that data type.

Note:

Functions generated using `inverseMap` will never be as efficient as their hand-written counterparts. This is because they need to perform dynamic lookups to get a return value. Even worse, the inverse of `t => a` has O(n) worst-case time complexity if we don't have a comparator function for `a` (we are forced to use an associated list/array for lookups). Therefore, I made a second function, `inverseMapWithComparator`, which takes a `compare` function for `a` and uses a `Map.t(a, t)` for the internal lookup table, which should be roughly O(log(n)).

I also included some tests for these functions.

### CODE REVIEW

File: src/extensions/Relude_Extensions_BoundedEnum.re
Comment (andywhite37) [inline src/extensions/Relude_Extensions_BoundedEnum.re:81]: Not 100% sure, but this function or something like it might already exist in https://github.com/reazen/relude/blob/master/src/extensions/Relude_Extensions_Enum.re - like a function to enumerate all the values of an `Enum`.

Also a tiny nitpick, but it might be best to not use `return` as the name of a variable - that threw me off at first glance.

Comment (austindd) [reply src/extensions/Relude_Extensions_BoundedEnum.re:81]: That function from `Enum` that enumerates all the values from an enum might be interchangeable here, I agree. I'll take a look and see if I can just reuse that. But I am not totally sure, because this needs to construct not just a list of `E.t`, but an associative list of `('a, E.t)`.

One reason for the current design was to keep it efficient using a tail-recursive loop, and the most efficient version of that is to return directly from one of the branches of the `switch` expression, so that there are no operations left after the recursive loop. This requires constructing the closure inside the `switch` expression directly. Although I'm not sure it's that important, and failing to reuse an existing function could result in larger bundle/binary sizes.

I think it's probably less efficient to reuse the existing function and then `zip` it into a new associative list, only to convert it into an array afterwards. Seems like a lot of allocations, IMO. But I don't have strong opinions about it.

Comment (andywhite37) [reply src/extensions/Relude_Extensions_BoundedEnum.re:81]: My point is that I'd rather put a particular algorithm in a single place and reuse it, rather than having to nest these helper functions inside other functions. If we did that all over the place, we'd have lots of implementations of every particular function to keep track of, rather than a single implementation, which is reused everywhere. If your `loop` function does something useful in the context of `Enum` or `BoundedEnum`, I would say it would be useful to add it as a new helper function to expose at the top-level and reuse it in these other functions. If your `loop` function already exists in concept somewhere else, but is a non-ideal implementation, I'd rather just reuse that function anyway, and put the onus on that reusable function to be improved. I'm not saying we need to blindly reuse functions and force ourselves to do additional allocations or extra big-O complexity - but just that we should create solid top-level functions and reuse them where appropriate, rather than re-inventing or re-working algorithms in the contexts of other functions.

File: src/extensions/Relude_Extensions_BoundedEnum.re
Comment (austindd) [reply src/extensions/Relude_Extensions_BoundedEnum.re:145]: Right, I remember wanting to use that function, but then running into cyclic dependency issues, since `Relude_Ordering` includes a functor for `BoundedEnumExtensions`. Hence why I even used the raw `BsBastet.Interface.ordering` type instead of Relude's own type alias. I'm open to alternatives, though.

File: src/extensions/Relude_Extensions_BoundedEnum.re
Comment (austindd) [reply src/extensions/Relude_Extensions_BoundedEnum.re:81]: After a second look, I think it might work out nicely to reuse the function from Enum Extensions.

Comment (andywhite37) [reply src/extensions/Relude_Extensions_BoundedEnum.re:81]: Cool - sounds good. If you think there isn't a good/workable thing to reuse that is totally fair too - I think in many cases it makes sense to create super-specialized versions of functions for the sake of performance, but I just don't want to default to doing that first if we can.

File: src/extensions/Relude_Extensions_BoundedEnum.re
Comment (andywhite37) [reply src/extensions/Relude_Extensions_BoundedEnum.re:145]: Ah, good point... in that case, it's probably not worth worrying about

**Comment by andywhite37** [comment]
@austindd - I don't want to hold up your progress on this since it's working now. If you would like, we can just merge this as-is, and if it makes sense to do any follow-on work, that's cool. It's not that big a deal either way.

**Comment by austindd** [comment]
@andywhite37, Sorry, I meant to tackle this earlier, but just got extremely busy recently. I actually do want to finish it up in the next few days or so, but if you would rather merge it now, then that’s cool too.

**Comment by andywhite37** [comment]
I don't need to merge it now - I just didn't want you to feel discouraged. I'm happy to wait for whatever additional changes you want to make.

**Comment by austindd** [comment]
@andywhite37 @mlms13 @johnhaley81 I just pushed some changes to the repo. It's not quite ready to merge. Just wanted to get your input on some alternative implementations of `inverseMapEqBy` and `inverseMapOrdBy`, which reuse some existing functions from `EnumExtensions`. These new versions are named `inverseMapEqBy2` and `inverseMapOrdBy2`. The old versions are left in for you to compare.

My thoughts:

1. Reusing functions from `EnumExtensions` means that we must apply the `EnumExtensions` functor inside `BoundedEnumExtensions` so we can access the resulting functions. I'm not sure if this is desirable, because it wasn't included previously. I guess it depends on how modular you want it to be.
2. This does reduce the Reason code size by a small amount. The JS output is reduced a bit more, but including other extension modules can potentially increase the JS bundle size, so take it with a grain of salt.
3. As @andywhite37 said, this does reduce the cognitive load, at least for `inverseMapEqBy`, but I'm not convinced that it's that much simpler for `inverseMapOrdBy`. Also, the other functions are already implemented with code reuse in mind.

**Comment by johnhaley81** [comment]
Personally, I find the V2 implementations easier to follow and thought that the data structures more properly modeled what was being done. I also like that it was reusing the stuff form `EnumExtensions`. I don't think that `O(n)` performance here is really something to worry about since I would imagine that most enum options would be pretty small relatively speaking.

I'd vote for the V2 implementations.

File: src/extensions/Relude_Extensions_BoundedEnum.re
Comment (andywhite37) [inline src/extensions/Relude_Extensions_BoundedEnum.re:96]: My suggestion to reuse existing functions doesn't necessarily mean you have to reuse functions that don't make sense. Like if you're wanting to operate on `Array` vs. `List`, it would probably make sense to create a `upFromIncludingAsArray` function and use that, rather than using a function that operates on `List`, then mapping the `List`, then converting it to an `Array`. That's probably killing any micro-optimizations you might be gaining by using `mapU`/`reduceU` below.

Basically the idea is to reuse (or create) reusable functions that do something and then use them in places that need that thing. I'd prefer not to see lots of functions like this that have inline definitions of functions like `find`.

File: src/extensions/Relude_Extensions_BoundedEnum.re
Comment (andywhite37) [inline src/extensions/Relude_Extensions_BoundedEnum.re:112]: Maybe I missed this before, but isn't this just the normal `Array` `find` function? I'm not sure, but I suspect this already exists on `Array` (of if not, it could be added), or maybe on `Belt.Array`. Does it need to be defined inline like this?

File: src/extensions/Relude_Extensions_BoundedEnum.re
Comment (andywhite37) [inline src/extensions/Relude_Extensions_BoundedEnum.re:215]: Curious why `reduceU` over just the normal `foldLeft` or `reduce` on `List`? We don't really use the uncurried functions anywhere else in this library, so not sure it makes sense to deviate from that convention for this.

File: src/extensions/Relude_Extensions_BoundedEnum.re
Comment (andywhite37) [inline src/extensions/Relude_Extensions_BoundedEnum.re:191]: I wonder if there's a function somewhere that does this so it doesn't need to be defined inline like this. It seems to be reducing an iterable type of thing into a `Map`.

**Review by andywhite37 — CHANGES_REQUESTED** [review]
I think I prefer the v2 ones that reuse existing functions. However, I think there are maybe some more things that could potentially be reused from other places. Basically just wanting to not redefine existing functions inline inside these new functions.

File: src/extensions/Relude_Extensions_BoundedEnum.re
Comment (austindd) [reply src/extensions/Relude_Extensions_BoundedEnum.re:191]: I can work on simplifying this, for sure. There is definitely a fold-like abstraction here.

File: src/extensions/Relude_Extensions_BoundedEnum.re
Comment (austindd) [reply src/extensions/Relude_Extensions_BoundedEnum.re:215]: Ah, I wasn't aware of that particular convention. Just reduces the overhead of autocurrying function calls. In the future, ideally, autocurrying could potentially be eliminated using third-party static analysis/optimization tools anyway... I'll go ahead and change it.

File: src/extensions/Relude_Extensions_BoundedEnum.re
Comment (austindd) [reply src/extensions/Relude_Extensions_BoundedEnum.re:112]: Yeah, I think you're right. I'll revisit that.

**Comment by andywhite37** [comment]
I checked out @austindd 's branch and did a little follow-on work to get these useful functions merged! Thanks @austindd for getting the ball rolling on these. Feel free to PR further changes to them if I didn't capture the intent, or didn't implement as you intended - the tests still seemed to pass, so I think it's at least correct in that sense.

**Comment by austindd** [comment]
@andywhite37 I appreciate the help in getting it finished. I apologize for letting this sit here. Just have a lot going on in my life right now (mostly good things), so I haven't had much time for open-source. The functions look good to me. I'm happy with it!

**Comment by andywhite37** [comment]
No worries! I had kind of forgotten about it, and we got another PR in and I remembered, so I figured I'd try to just get it in there while I was back in the code. Thanks - I like that function, it's not something I was aware of, but that's a nice utility.

**Pull request merged**
